src/ObservationDims.jl

```diff
@@ -1,5 +1,159 @@
 module ObservationDims
 
-greet() = print("Hello World!")
+using NamedDims
+using AxisArrays
+using Distributions
+
+export obs_arrangement, organise_obs
+export SingleObs, IteratorOfObs, ArraySlicesofObs
+export MatrixRowsOfObs, MatrixColsOfObs
+
+"""
+    type ObsArrangement
+
+Defines the orientation of data that is expected by a function
+"""
+abstract type ObsArrangement end
+
+struct SingleObs <: ObsArrangement end
+struct IteratorOfObs <: ObsArrangement end
+struct ArraySlicesOfObs{D} <: ObsArrangement end
+
+const MatrixRowsOfObs = ArraySlicesOfObs{1}
+const MatrixColsOfObs = ArraySlicesOfObs{2}
+
+
+"""
+    organise_obs(f, data; obsdim=nothing)
+
+Organise the `data` according to the `ObsArrangement` expected by some function `f`.
+"""
+function organise_obs(f, data; obsdim=nothing)
+    return organise_obs(obs_arrangement(f), data; obsdim=obsdim)
+end
+
+
+# Specify arrangement based on type of data:
+
+# Scalars have no "orientation" so no rearrangement required
+for T in (Sampleable, Number, Symbol)
+    @eval organise_obs(f, data::$T; obsdim=nothing) = data
+    @eval organise_obs(::SingleObs, data::$T; obsdim=nothing) = data
+end
+
+## Vectors: obsdim is optional, we may or may not need it.
+for T in (Any, AbstractVector)
+
+    # Iterator -> IteratorOfObs
+    @eval organise_obs(::IteratorOfObs, obs_iter::$T; obsdim=nothing) = obs_iter
+
+    # Iterator -> ArraySlicesOfObs
+    @eval function organise_obs(::ArraySlicesOfObs{D}, obs_iter::$T; obsdim=nothing) where D
+
+        # we assume all obs have same number of dimensions else nothing makes sense
+        ndims_per_obs = ndims(first(obs_iter))
+
+        # If collection of scalars then return the collection
+        if ndims_per_obs == 0
+            return collect(obs_iter)
+        end
+
+        # This should just be a mapreduce but that is slow
+        # see https://github.com/JuliaLang/julia/issues/31137
+        shaped_obs = Base.Generator(obs_iter) do obs
+            new_shape = ntuple(ndims_per_obs + 1) do ii
+                if ii < D
+                    size(obs, ii)
+                elseif ii > D
+                    size(obs, ii-1)
+                else  # ii = D
+                    1
+                end
+            end
+            # add singleton dim that we wil concatenate on
+            Base.ReshapedArray(obs, new_shape, ())
+        end
+        return cat(shaped_obs...; dims=D)
+    end
+end
+
+# Specify arrangement based on desired ObsArrangement:
+
+# Any -> SingleObs: never any need to rearrage
+organise_obs(::SingleObs, data; obsdim=nothing) = data
+
+# If data is an array we can safely drop single length dimensions
+# This lets us compute metrics on pairs of objects that are (3,) and (3, 1)
+function organise_obs(::SingleObs, data::AbstractArray; obsdim=nothing)
+    to_drop = findall(size(data) .== 1)
+    data = isempty(to_drop) ? data : dropdims(data; dims=Tuple(to_drop))
+    return data
+end
```

Comment:
We should open an issue about this, since it's quite opinionated, and not sure whether or not we want it

Reply:
i'm leaning towards not wanting it. it's kinda weird.

Reply:
what in particular?

Reply:
Automatically dropping 1-length dimensions which amounts to reducing (2, 1) -> (2,) or (3, 1, 5) -> (3, 5).  I agree it's presumptuous as we have our own reasons for doing it.

Reply:
https://github.com/invenia/ObservationDims.jl/issues/10